File: Part3/customerPart3.py
import psycopg2
import pandas as pd
import json
from datetime import datetime

def GetCustomerInfo(id, startDate=None, endDate=None):
    dataTable = pd.DataFrame()

    try:
        with open('appsettings.json', 'r') as f:
            config = json.load(f)
        db_config = config["ConnectionStrings"]["DefaultConnection"]

        with psycopg2.connect(
            host=db_config["host"],
            database=db_config["database"],
            user=db_config["user"],
            password=db_config["password"]
        ) as conn:
            conn.autocommit = True

            query = "SELECT id, name, created_at FROM Customer WHERE id = %s"
            params = [id]

            if not startDate:
                startDate = "1950-01-01"
            if not endDate:
                endDate = datetime.now().strftime('%Y-%m-%d')
            query += " AND created_at BETWEEN %s AND %s"
            params.extend([startDate, endDate])

            data = pd.read_sql(query, conn, params=params)

            dataTable = data.astype({
                "id": str,
                "name": str
            })

    except Exception as e:
        print(f"error: {e}")

    return dataTable

File: Part3/customerPart3.cs
using System.Data;
using Microsoft.Extensions.Configuration;
using Newtonsoft.Json.Linq;
using System.Data.SqlClient;

public DataTable GetCustomerInfo(string id, DateTime? startDate = null, DateTime? endDate = null)
{
    var dt = new DataTable();

    var config = JsonConvert.DeserializeObject<JObject>(File.ReadAllText("appsettings.json"));
    var dbConfig = config["ConnectionStrings"]["DefaultConnection"];

    using (var conn = new SqlConnection(
        $"Server={dbConfig["host"]};Database={dbConfig["database"]};User Id={dbConfig["user"]};Password={dbConfig["password"]};"))
    {
        conn.Open();

        string sql = "SELECT id, name, email FROM Customer WHERE id = @id";
        var cmd = new SqlCommand();
        cmd.Connection = conn;
        cmd.Parameters.AddWithValue("@id", id);

        if (!startDate.HasValue)
            startDate = new DateTime(1950, 1, 1); 

        if (!endDate.HasValue)
            endDate = DateTime.Now; 

        sql += " AND created_at BETWEEN @start AND @end";
        cmd.Parameters.Add("@start", SqlDbType.DateTime).Value = startDate.Value;
        cmd.Parameters.Add("@end", SqlDbType.DateTime).Value = endDate.Value;

        cmd.CommandText = sql;

        using (var da = new SqlDataAdapter(cmd))
        {
            da.Fill(dt);
        }
    }

    return dt;
}
